### ISSUE TITLE
Help with implementing circuit breaker with Async API

### ISSUE THREAD

**Issue by li-darren** [resilience4j/resilience4j#1256]
Hi,

Sorry if this is not the right place to comment, but I was taking a look at this below, and I'm not sure if it solves the problem I have. I also do not know how to mark this as a question.
https://github.com/resilience4j/resilience4j/issues/823 

I have an async publishing API that I want to use a circuit breaker and retries with. I looked at some of the methods used with creating Thread Pools and BulkHeads, but does that make sense in this case? The API is async and will send a callback back indicating success or not, so I'm not sure if we need another thread pool to handle sending off messages to the publisher. 

Basically, I am looking for something along the lines of this, but for an async method. 
https://github.com/resilience4j/resilience4j/blob/6e5a7b98b3aacfd64c837706893d04b8ba8ef298/resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java#L140

Do you know how I can approach this issue?

Thanks,

**Comment by RobWin** [comment]
Hi,
do you use any specific framework like Spring Boot?

I need more context, since there are different strategies. Do you control the method/function which returns a `CompletableFuture`? 

You can either decorate the inner `Supplier` function:

```
Supplier<String> decoratedSupplier = Decorators
    .ofSupplier(() -> yourService.yourMethod(param1, param2))
    .withCircuitBreaker(circuitBreaker)
    .withRetry(retry)
    .decorate();
        
CompletableFuture<String> future = CompletableFuture.supplyAsync(decoratedSupplier);
```

Or you can run the retry attempts in a different thread asynchronously. But in that case you need a `ScheduledExecutorService` to schedule retry attempts.

```
CompletableFuture<String> completionStage = Decorators
    .ofCompletionStage(() -> CompletableFuture.supplyAsync(() -> yourService.yourMethod(param1, param2))
    .withCircuitBreaker(circuitBreaker)
    .withRetry(retry, Executors.newScheduledThreadPool(10))
    .get().toCompletableFuture();
```

**Comment by li-darren** [comment]
Thanks for the reply. I am using Spring Boot. The API I am calling is a void method, so it is almost like a send it and forget it kind of thing, I'm not sure anything returns a completable future. It returns a callback though. I guess it doesn't hurt to just share, I am using an async Kafka producer.
I guess it may be possible to wrap the function like this to get a completable future?
https://github.com/1and1/reactive/blob/e582c0bdbfb4ab2a0780c77419d0d3ee67f08067/reactive-kafka/src/main/java/net/oneandone/reactive/kafka/CompletableKafkaProducer.java#L42

"The Kafka Producer has a send() method which is asynchronous. Calling the send method adds the record to the output buffer and return right away. The buffer is used to batch records for efficient IO and compression. The Kafka Producer configures acks to control record durability. The ”all” acks setting ensures full commit of record to all replicas and is most durable and least fast setting. The Kafka Producer can automatically retry failed requests. The Producer has buffers of unsent records per topic partition (sized at batch.size)."
http://cloudurable.com/blog/kafka-tutorial-kafka-producer-advanced-java-examples/index.html#:~:text=The%20Kafka%20Producer%20has%20a,acks%20to%20control%20record%20durability.

**Comment by RobWin** [comment]
You should never do retry on application-level when using a Kafka producer. Please use the Kafka client options to retry.

**Comment by li-darren** [comment]
Would it not be good to utilize the circuit breaker?

**Comment by RobWin** [comment]
Yes, a CircuitBreaker makes sense. If you use  our Spring Boot2 starter, you can also use Annotations on class or method level, when your class is a bean

**Comment by li-darren** [comment]
Yep, so that is what I am trying to configure. So i'm wondering, how should I configure the circuit breaker in this case with an async api? It seems like all of the circuit breaker code decorators is for synchronous methods. Would you recommend manually calling acquirepermission() and the onSuccess(...) and onError(...) methods? Or is there some better way to do this?

**Comment by RobWin** [comment]
Thr annotations also work on methods which return a CompletableFuture. See your Spring Boot 2 demo

https://github.com/resilience4j/resilience4j-spring-boot2-demo/blob/master/src/main/java/io/github/robwin/service/BackendAService.java#L126

**Comment by li-darren** [comment]
Does it make sense to use another Thread pool to deal with sending messages to the API, when Kafka producer has it's own thread pool in it's implementation? In addition, how would resilience4j know whether the API call was successful or not? Does it call get on the completablefuture?

**Comment by RobWin** [comment]
No, I think another ThreadPool doesn't make sense, if you monitor your existing one.
If the CompletableFuture is completed successfully, the CB records as a success.
If the future is completed exceptionally, the CB checks the exception and based on your configuration either records the exception as a failure, as a success or ingores it.

**Comment by li-darren** [comment]
Hey I'm a bit stuck again. I tried wrapping the method

**Comment by li-darren** [comment]
![image](https://user-images.githubusercontent.com/43430567/101700343-dfb36900-3a4a-11eb-862a-f5bfee196211.png)
Hey I'm a bit stuck again...

In the function that returns a Completable Future, I declare a completable future at the top, and through the Kafka callback, I either complete() or completeExceptionally(), depending on if the callback returns an exception.

In the circuit breaker code above, it seems like the calls are stuck on the buffer, increasing the getNumberOfBufferedCalls and getNumberOfSuccessfulCalls up the the max buffered size in the configuration. Do you know why it may be the case that the calls are left buffered?

**Comment by li-darren** [comment]
In addition, won't the supplier always be able to get the completable future? So that means the completedExceptionally case will never happen...
![image](https://user-images.githubusercontent.com/43430567/101703949-b0ecc100-3a51-11eb-815e-4696969a2deb.png)

**Comment by RobWin** [comment]
See my code example above.
If your method already returns a `CompletableFuture`, you must not use `Decorators.ofSupplier`, but instead

```
CompletableFuture<String> completionStage = Decorators
    .ofCompletionStage(() -> funcThatReturnsCompletableFuture)
    .withCircuitBreaker(circuitBreaker)
    .get().toCompletableFuture();
```

**Comment by Romeh** [comment]
closing this ticket due to inactivity , please feel free to open a new one if you still have questions

**Comment by roshansinghkande** [comment]
@RobWin  It's very kind of you to help users establish good usage patterns like above. This is massively helpful. Thanks!
